### PR TITLE
Add the `timestamp without time zone` data type

### PR DESCRIPTION
**Pull request by kovrus**
## Summary of the changes / Why this improves CrateDB

The PR introduces the `timestamp without time zone` data type.

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/protocols/postgres/types/TimestampZType.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+package io.crate.protocols.postgres.types;
+
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import javax.annotation.Nonnull;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
+
+final class TimestampZType extends BaseTimestampType {
```
Comment (kovrus) [inline sql/src/main/java/io/crate/protocols/postgres/types/TimestampZType.java]: I keep it as it was before, the file was renamed and pointed to the abstract class. It still uses joda time for formatters, would suggest to change it in the separate PR.

File: sql/src/test/java/io/crate/protocols/postgres/types/TimestampZTypeStringDecodeTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import org.junit.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Locale;
+
+import static com.carrotsearch.randomizedtesting.RandomizedTest.$;
+import static org.hamcrest.Matchers.is;
+
+public class TimestampZTypeStringDecodeTest extends BasePGTypeTest<Long> {
```
Comment (kovrus) [inline sql/src/test/java/io/crate/protocols/postgres/types/TimestampZTypeStringDecodeTest.java]: It probably should not be reviewed, it has the old code, just the class was renamed.

File: sql/src/test/java/io/crate/protocols/postgres/types/TimestampZTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.base.Charsets;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.junit.Test;
+
+import java.nio.charset.StandardCharsets;
+
+import static org.hamcrest.Matchers.is;
+
+public class TimestampZTypeTest extends BasePGTypeTest<Long> {
```
Comment (kovrus) [inline sql/src/test/java/io/crate/protocols/postgres/types/TimestampZTypeTest.java]: Same here. Only the test class was renamed.

File: common/src/main/java/io/crate/types/BaseTimestampType.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    public Precedence precedence() {
+        return Precedence.LongType;
+    }
+
+    @Override
+    public Streamer<Long> streamer() {
+        return this;
+    }
+
+    public abstract Long valueFrom(String timestamp);
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/BaseTimestampType.java]:
```suggestion
    public abstract long valueFrom(String timestamp);
```

File: common/src/main/java/io/crate/types/BaseTimestampType.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+public abstract class BaseTimestampType
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/BaseTimestampType.java]: Have you considered a more compositional approach? E.g. having a `TimestampType` that takes a `Function<String, Long> parse` argument and then instantiate the two types accordingly?

Has the advantage that the parse method could be more easily tested in isolution and that there are no bi-directional calls from child<->parent.

File: es/es-server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 6]
```diff
@@ -335,18 +345,21 @@ public Object valueForDisplay(Object value) {
     }
 
     private Explicit<Boolean> ignoreMalformed;
+    private Boolean ignoreTimezone;
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java]: could this be final?

File: common/src/main/java/io/crate/types/TimestampType.java [hunk 7]
```diff
@@ -29,44 +35,115 @@
...
 import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
 import static java.time.format.DateTimeFormatter.ISO_LOCAL_TIME;
 
-public final class TimestampType extends BaseTimestampType {
+public final class TimestampType extends DataType<Long>
+    implements FixedWidthType, Streamer<Long> {
 
-    public static final TimestampType INSTANCE = new TimestampType();
-    public static final int ID = 15;
+    public static final int ID_WITH_TZ = 15;
```
Comment (kovrus) [inline common/src/main/java/io/crate/types/TimestampType.java]: looks a bit weird with these constants, but we are forced to have them when we decide on data type based on its in in switch statements

File: common/src/main/java/io/crate/types/TimestampType.java [hunk 8]
```diff
@@ -29,44 +35,115 @@
...
-    private static final DateTimeFormatter PARSER = new DateTimeFormatterBuilder()
-        .parseCaseInsensitive()
-        .append(ISO_LOCAL_DATE)
-        .optionalStart()
-            .appendLiteral('T')
-            .append(ISO_LOCAL_TIME)
-            .optionalStart()
-                .appendPattern("[VV][x][xx][xxx]")
-        .toFormatter(Locale.ENGLISH).withResolverStyle(ResolverStyle.STRICT);
+    public static final TimestampType WITH_TZ = new TimestampType(
```
Comment (smakalias) [inline common/src/main/java/io/crate/types/TimestampType.java]: I think I would also add the `INSTANCE` here (and below) to make it explicit that they're actually timestamp instances.
It seems more clear when using them.

What do you think?

File: common/src/main/java/io/crate/types/TimestampType.java (hunk 7 above)
Comment (smakalias) [reply common/src/main/java/io/crate/types/TimestampType.java]: Would it be better if we keep them private (hide them) and use instance.id() instead?

Comment (mfussenegger) [reply common/src/main/java/io/crate/types/TimestampType.java]: `switch` requires constants and doesn't work on `id()`. 

I don't see a problem exposing them like this.

File: common/src/main/java/io/crate/types/TimestampType.java [hunk 9]
```diff
@@ -108,4 +130,38 @@ public void writeValueTo(StreamOutput out, Long v) throws IOException {
...
+
+    static long parseTimestamp(String timestamp) {
+        try {
+            return Long.valueOf(timestamp);
+        } catch (NumberFormatException e) {
+            return TimestampFormat.parseTimestampString(timestamp);
+        }
+    }
+
+    static long parseTimestampIgnoreTimeZone(String timestamp) {
```
Comment (smakalias) [inline common/src/main/java/io/crate/types/TimestampType.java]:
```suggestion
    static long parseTimestampWithOptionalTimeZone(String timestamp) {
```

Comment (kovrus) [reply common/src/main/java/io/crate/types/TimestampType.java]: it is not optional, it ignores it. It parse it with or without time zone, but then converts it to UTC ignoring the time zone information.

Comment (smakalias) [reply common/src/main/java/io/crate/types/TimestampType.java]: ah, no.. I read it wrong. It's fine as it is.

**Pull request merged**